### PR TITLE
[webpack] Localization plugin prototype.

### PR DESCRIPTION
**Pull request by iclanton**

### CODE REVIEW

File: common/changes/@microsoft/rush/ianc-loc-plugin-prototype_2019-05-13-07-53.json [hunk 1]
```diff
@@ -1,11 +1,11 @@
 {
```
Comment (halfnibble) [inline common/changes/@microsoft/rush/ianc-loc-plugin-prototype_2019-05-13-07-53.json]: Was this accidentally carried over from the `rush list --json` PR?

Comment (iclanton) [reply common/changes/@microsoft/rush/ianc-loc-plugin-prototype_2019-05-13-07-53.json]: Oh that's weird. Should go away after Rush is published.

File: build-tests/localization-plugin-test/build.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
+const fsx = require('fs-extra');
```
Comment (apostolisms) [inline build-tests/localization-plugin-test/build.js]: do we want to use fs-extra?

Comment (iclanton) [reply build-tests/localization-plugin-test/build.js]: No - I'll fix that.

Comment (halfnibble) [reply build-tests/localization-plugin-test/build.js]: Are you gonna use the FileDeletion utility I thought should be extracted to node-core-library?

File: webpack/localization-plugin/src/loaders/SingleLocaleLoader.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  passthroughLocale: boolean;
+}
+
+export default function (this: loader.LoaderContext, content: string): string {
+  const {
+    resolvedStrings,
+    passthroughLocale
+  } = loaderUtils.getOptions(this) as ILocJsonLoaderOptions;
+  const locJsonFilePath: string = this.resourcePath;
+  const normalizedLocJsonFilePath: string = locJsonFilePath.toUpperCase();
```
Comment (apostolisms) [inline webpack/localization-plugin/src/loaders/SingleLocaleLoader.ts]: Any particular reason why UpperCase ?

File: build-tests/localization-plugin-test/build.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+const child_process = require('child_process');
+const path = require('path');
+const process = require('process');
+
+const { LocJsonPreprocessor } = require('@rushstack/localization-plugin');
+
+function executeCommand(command) {
+  console.log('---> ' + command);
```
[the remark this answers is not in the document]
Comment (iclanton) [reply build-tests/localization-plugin-test/build.js]: yeah these are in the test project.

File: webpack/localization-plugin/src/LocalizationPlugin.ts [hunk 5]
```diff
@@ -0,0 +1,732 @@
...
+ */
+export interface IStringPlaceholder {
+  value: string;
+  suffix: string;
+}
+
+const PLUGIN_NAME: string = 'localization';
+const LOCALE_FILENAME_PLACEHOLDER: string = '[locale]';
+const LOCALE_FILENAME_PLACEHOLDER_REGEX: RegExp = new RegExp(lodash.escapeRegExp(LOCALE_FILENAME_PLACEHOLDER), 'g');
+const STRING_PLACEHOLDER_PREFIX: string = '-LOCALIZED-STRING-f12dy0i7-n4bo-dqwj-39gf-sasqehjmihz9';
```
Comment (halfnibble) [inline webpack/localization-plugin/src/LocalizationPlugin.ts]: Did you intend for this to start with a dash `-`?

Comment (iclanton) [reply webpack/localization-plugin/src/LocalizationPlugin.ts]: Yeah.

File: webpack/localization-plugin/src/LocalizationPlugin.ts [hunk 6]
```diff
@@ -0,0 +1,732 @@
...
+  private _serveLocale: string;
+  private _usePassthroughForServe: boolean | undefined;
+  /**
+   * The outermost map's keys are the locale names.
+   * The middle map's keys are the resolved, uppercased file names.
+   * The innermost map's keys are the string identifiers and its values are the string values.
+   */
+  private _resolvedLocalizedStrings: Map<string, Map<string, Map<string, string>>>;
+
+  public constructor(options: ILocalizationPluginOptions) {
```
Comment (halfnibble) [inline webpack/localization-plugin/src/LocalizationPlugin.ts]: How will this work in our projects? Will we need to include large `ILocalizationPluginOptions` settings in the webpack config, or will this be generated somehow?

Also, if I modify a `.loc.json` file, will I need to amend the config and restart webpack?

Comment (iclanton) [reply webpack/localization-plugin/src/LocalizationPlugin.ts]: For the time being, yes. Right now this is oversimplified, but I want that to be improved soon.

File: webpack/localization-plugin/src/loaders/SingleLocaleLoader.ts (hunk 3 above)
Comment (iclanton) [reply webpack/localization-plugin/src/loaders/SingleLocaleLoader.ts]: That was to do case insensitive path comparison. Doesn't seem to matter, though. Removed.

**Pull request merged**